### ISSUE TITLE
[Feature Request] Import faSvg icons used by vuetify

### ISSUE THREAD

**Issue by P4sca1** [vuetifyjs/vuetify#9995]
### Problem to solve
Even though `faSvg` support landed in `2.1.15`, it is not quite usable yet.
Right now the docs tell you to do the following: `library.add(fas)`.
This causes 2 major problems:

1. Vuetify uses some far icons which are not being added to the library.
2. Adding all `fas` and `far` icons to the library causes too much overhead in page size.

Also `faSvg` is missing in the `Iconfont` type in `types/services/icons.d.ts` which causes typescript build to fail when using this font.

### Proposed solution
Import only the icons used by vuetify.
I managed to come up with this custom solution, but would be great if it would be built into vuetify.

```ts
...
import { VuetifyIcons } from 'vuetify/types/services/icons'
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'
import { library } from '@fortawesome/fontawesome-svg-core'

import {
  faCheck as fasCheck,
  faTimesCircle as fasTimesCircle,
  faTimes as fasTimes,
  faCheckCircle as fasCheckCircle,
  faInfoCircle as fasInfoCircle,
  faExclamation as fasExclamation,
  faExclamationTriangle as fasExclamationTriangle,
  faChevronLeft as fasChevronLeft,
  faChevronRight as fasChevronRight,
  faCheckSquare as fasCheckSquare,
  faMinusSquare as fasMinusSquare,
  faCircle as fasCircle,
  faSortUp as fasSortUp,
  faChevronDown as fasChevronDown,
  faBars as fasBars,
  faCaretDown as fasCaretDown,
  faEdit as fasEdit,
  faStar as fasStar,
  faStarHalf as fasStarHalf,
  faSync as fasSync,
  faStepBackward as fasStepBackward,
  faStepForward as fasStepForward,
  faArrowsAltV as fasArrowsAltV,
  faPaperclip as fasPaperclip,
  IconDefinition,
} from '@fortawesome/free-solid-svg-icons'

import {
  faSquare as farSquare,
  faDotCircle as farDotCircle,
  faCircle as farCircle,
  faStar as farStar,
} from '@fortawesome/free-regular-svg-icons'

const makeIconComponentDeclaration = (icon: IconDefinition) => {
  library.add(icon) // This is important!

  return {
    component: FontAwesomeIcon,
    props: {
      icon: [icon.prefix, icon.iconName],
    },
  }
}

// Maps internal Vuetify icon names to Font-Awesome 5+ icon names.
const ICONS_FONTAWESOME: VuetifyIcons = {
  complete: makeIconComponentDeclaration(fasCheck),
  cancel: makeIconComponentDeclaration(fasTimesCircle),
  close: makeIconComponentDeclaration(fasTimes),
  delete: makeIconComponentDeclaration(fasTimesCircle), // delete (e.g. v-chip close)
  clear: makeIconComponentDeclaration(fasTimesCircle), // delete (e.g. v-chip close)
  success: makeIconComponentDeclaration(fasCheckCircle),
  info: makeIconComponentDeclaration(fasInfoCircle),
  warning: makeIconComponentDeclaration(fasExclamation),
  error: makeIconComponentDeclaration(fasExclamationTriangle),
  prev: makeIconComponentDeclaration(fasChevronLeft),
  next: makeIconComponentDeclaration(fasChevronRight),
  checkboxOn: makeIconComponentDeclaration(fasCheckSquare),
  checkboxOff: makeIconComponentDeclaration(farSquare), // note 'far'
  checkboxIndeterminate: makeIconComponentDeclaration(fasMinusSquare),
  delimiter: makeIconComponentDeclaration(fasCircle), // for carousel
  sort: makeIconComponentDeclaration(fasSortUp),
  expand: makeIconComponentDeclaration(fasChevronDown),
  menu: makeIconComponentDeclaration(fasBars),
  subgroup: makeIconComponentDeclaration(fasCaretDown),
  dropdown: makeIconComponentDeclaration(fasCaretDown),
  radioOn: makeIconComponentDeclaration(farDotCircle),
  radioOff: makeIconComponentDeclaration(farCircle),
  edit: makeIconComponentDeclaration(fasEdit),
  ratingEmpty: makeIconComponentDeclaration(farStar),
  ratingFull: makeIconComponentDeclaration(fasStar),
  ratingHalf: makeIconComponentDeclaration(fasStarHalf),
  loading: makeIconComponentDeclaration(fasSync),
  first: makeIconComponentDeclaration(fasStepBackward),
  last: makeIconComponentDeclaration(fasStepForward),
  unfold: makeIconComponentDeclaration(fasArrowsAltV),
  file: makeIconComponentDeclaration(fasPaperclip),
}

...

new Vuetify({
  ...
  icons: {
    // Little hack to support font awesome svg icons.
    // Progress: https://github.com/vuetifyjs/vuetify/pull/8092
    iconfont: 'faSvg' as 'fa',
    values: ICONS_FONTAWESOME,
  },
})

...
```

Also add `faSvg` to the `Iconfont` type like so:
`export type Iconfont = 'mdi' | 'mdiSvg' | 'md' | 'fa' | 'fa4' | 'faSvg'`

**Comment by basselAhmed** [comment]
@P4sca1 I don't think it's a good idea to have the icons built into Vuetify, this way it leaves it up to the user to import only what's needed.

> Also faSvg is missing in the Iconfont type in types/services/icons.d.ts which causes typescript build to fail when using this font.
I had that type defined correctly in my PR originally but I guess when my branch was updated with dev branch some changes were overwritten.
Will make another PR to fix this I guess.

**Comment by P4sca1** [comment]
@basselAhmed Alright, thank you.
I guess we should at least update the docs with an example on how to import only the needed icons.
Right now the docs are showing bad practices and as I said some icons are missing.

**Comment by basselAhmed** [comment]
@P4sca1  I've created a PR for the bug fix but no documentation updates yet, but I definitely agree with you in that regard.

**Comment by P4sca1** [comment]
Great, thank you for your effort!

**Comment by MajesticPotatoe** [comment]
Would have to agree with @basselAhmed on this. We have a no-dependency policy anyways. Offering it for faSvg would mean we would need to offer it for other libraries, which would add unnecessary bloat, when simply a user can define it themselves.

As for the icon docs, there are many examples of importing icons, though not very easy to find, there is one that imports just what you need under fa pro icons https://vuetifyjs.com/en/customization/icons#custom-font-awesome-pro-icons, though ultimately we already have much needed icon doc update issue, which would be better mentioned over in https://github.com/vuetifyjs/vuetify/issues/6033#issuecomment-450908147

Also pr for types have been added.

**Comment by P4sca1** [comment]
Alright. Because the type bug has already been adressed and icon documentation is discussed in another issue, I am going to close this issue.

**Comment by SirLamer** [comment]
Just to contribute to this a bit... I don't see how this proposal involves bundling font awesome with Vuetify. Supporting it is not the same as integration.

The OP's point remains a huge issue... it's not viable to ship with the entire fontawesome library embedded because it is massive, it is essential to pick from Font Awesome's library in order to use it. If this remains unaddressed then it might as well not be supported at all.

Others have proposed supporting a per-component object-based import solution like the implementation for the Material Design icon library and that seems like a good solution to me as it would lend toward shakedown very naturally, however the Vuetify SVG library bundles all of the icons within their own object package which is itself not suitable to being just passed in the main slot of the VIcon component, it would have to be passed as a component property.

**Comment by Tronix117** [comment]
I added the support for Font Awesome in the PR #10804 

I agree with @SirLamer , support is different as integration. Font Awesome Pro is one of the most popular icon library, and it makes sense to at least support it.

The implementation I did is fully retro-compatible and does not add much codes to it. It also add the ability to pass an SVG path with a different viewbox than the one VIcon uses by default. This is a generic feature.

The FA code, is just a mapping fonction with a check fonction, I don't think it adds unnecessary bloat, knowing that #9996 has already been merged, adding partial support.
